languageserver/src/main/scala/langserver/core/LspMessageReaderOperator.scala

```diff
@@ -0,0 +1,95 @@
+package langserver.core
+
+import langserver.messages.UnparsedLspMessage
+import monix.execution.Ack.Continue
+import monix.execution.{Ack, Scheduler}
+import monix.reactive.observables.ObservableLike.Operator
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.Future
+
+/*
+  Splits incoming bytes into unparsed LspMessages
+ */
+object LspMessageReaderOperator extends Operator[Array[Byte], List[UnparsedLspMessage]]{
+  //move this into LspMessage?
+  val CONTENT_LENGTH_KEY = "Content-Length"
+  def parseHeaders(input: String): Map[String, String] = {
+    val result = input.lines
+      .filter(_.nonEmpty)
+      .map(_.split(":"))
+      .map { case Array(k, v) => k -> v.trim }
+      .toMap
+    assert(result.contains(CONTENT_LENGTH_KEY), "Could not parse Content-Length from header")
+    result
+  }
+
+  sealed trait HeaderDelimiter {
```

Comment:
I chose to use a state machine here in case the header delimiter was broken up across multiple batches.  Not sure how I feel about the verbosity yet.

Reply:
I have a feeling we may be able to avoid this state machine by using fastparse-bytes http://www.lihaoyi.com/fastparse/#ByteParsers `Bytes ++ Bytes` and `Bytes.apply(index: Int)` are near constant. `.parse` takes an optional starting index which can be used to start parsing from somewhere inside the index.